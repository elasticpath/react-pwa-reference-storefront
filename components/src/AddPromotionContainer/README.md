# AddPromotionContainer

#### Description

Adds a promotion code to an order

#### Usage

```js
import { AddPromotionContainer } from '@elasticpath/store-components';
```

#### Example

```js
<AddPromotionContainer data={cartData} onSubmittedPromotion={() => { this.fetchCartData(); }} />
```

#### Properties

<!-- PROPS -->
