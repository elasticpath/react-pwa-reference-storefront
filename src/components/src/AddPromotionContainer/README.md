# AddPromotionContainer

#### Description

Apply coupons or other discounts to an order.

This component adds an **Add Promotion** button to a store page. When selected, the component changes to display an **Enter Coupon Code** field, **Apply Promotion** button, and a **Cancel** button.

- When a shopper enters a code and selects **Apply Promotion**, the component invokes the `onSubmittedPromotion` process. The process verifies the code and, if the code is valid, applies the discount to the order. If the code is not valid, an error is displayed.
- When a shopper selects **Cancel**, the expanded component collapses back to a button.

#### Usage

```js
import { AddPromotionContainer } from '@elasticpath/store-components';
```

#### Example

```js
<AddPromotionContainer data={cartData} onSubmittedPromotion={() => { this.fetchCartData(); }} />
```

#### Properties

<!-- PROPS -->
