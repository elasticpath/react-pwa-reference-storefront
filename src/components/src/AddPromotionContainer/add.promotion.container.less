/**
 * Copyright Â© 2019 Elastic Path Software Inc. All rights reserved.
 *
 * This is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this license. If not, see
 *
 *     https://www.gnu.org/licenses/
 *
 *
 */

 @import './../../../theme/common.less';

.add-promotion-container {
  .btn-add-promotion{
    text-transform: uppercase;
    margin: 5px 0;
    text-shadow: none;
    color: @secondComplimentColor;
    background-color: @mainBackgroundColor;
    border: solid 1px @secondComplimentColor;
    white-space: nowrap;
    min-width: 120px;
    height: 30px;
    font-size: 12px;
    font-weight: bold;

    &:hover {
      background-color: @secondComplimentColor;
      color: @mainBackgroundColor;
    }
  }
}
