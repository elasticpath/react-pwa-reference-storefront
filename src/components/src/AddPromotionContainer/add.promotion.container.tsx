/**
 * Copyright Â© 2019 Elastic Path Software Inc. All rights reserved.
 *
 * This is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this license. If not, see
 *
 *     https://www.gnu.org/licenses/
 *
 *
 */

import React, { Component } from 'react';
import intl from 'react-intl-universal';
import { login } from '../utils/AuthService';
import { cortexFetch } from '../utils/Cortex';
import { getConfig, IEpConfig } from '../utils/ConfigProvider';

import './add.promotion.container.less';

let Config: IEpConfig | any = {};

interface AddPromotionContainerProps {
  /** An array of orders for the shopper, where `string` is the URL to the cart data for an order. */
  data: {
    [key: string]: any
  },
  /** Verifies the coupon code and applies a valid coupon to the order. */
  onSubmittedPromotion: (...args: any[]) => any,
}

interface AddPromotionContainerState {
  isPromotionFormOpen: boolean,
  failedPromotion: boolean,
  promotionCode: string,
  couponFormLink: string,
}

class AddPromotionContainer extends Component<AddPromotionContainerProps, AddPromotionContainerState> {
  constructor(props) {
    super(props);
    const epConfig = getConfig();
    Config = epConfig.config;
    this.state = {
      isPromotionFormOpen: false,
      failedPromotion: false,
      promotionCode: '',
      couponFormLink: '',
    };
    this.setPromotionCode = this.setPromotionCode.bind(this);
    this.submitPromotionCode = this.submitPromotionCode.bind(this);
  }

  componentDidMount() {
    const { data } = this.props;
    if (data._order) {
      const couponFormUri = data._order[0]._couponinfo[0]._couponform[0].links.find(
        link => link.rel === 'applycouponaction',
      ).uri;
      this.setState({
        couponFormLink: couponFormUri,
      });
    }
  }

  setPromotionCode(event) {
    this.setState({ promotionCode: event.target.value });
  }

  openPromotionForm() {
    this.setState({
      isPromotionFormOpen: true,
    });
  }

  closePromotionForm() {
    this.setState({
      isPromotionFormOpen: false,
      failedPromotion: false,
    });
  }

  submitPromotionCode(event) {
    event.preventDefault();
    const { onSubmittedPromotion } = this.props;
    const { couponFormLink, promotionCode } = this.state;
    login().then(() => {
      cortexFetch(couponFormLink, {
        method: 'post',
        headers: {
          'Content-Type': 'application/json',
          Authorization: localStorage.getItem(`${Config.cortexApi.scope}_oAuthToken`),
        },
        body: JSON.stringify({
          code: promotionCode,
        }),
      }).then((res) => {
        if (res.status === 409) {
          this.setState({ failedPromotion: true });
        } else if (res.status === 201 || res.status === 200 || res.status === 204) {
          this.setState({ failedPromotion: false }, () => {
            this.closePromotionForm();
            onSubmittedPromotion();
          });
        }
      }).catch((error) => {
        // eslint-disable-next-line no-console
        console.error(error.message);
      });
    });
  }

  render() {
    const {
      isPromotionFormOpen,
      failedPromotion,
    } = this.state;
    if (isPromotionFormOpen) {
      return (
        <div className="add-promotion-container">
          <h2>
            {intl.get('add-promotion-title')}
          </h2>
          <form className="form-horizontal" onSubmit={this.submitPromotionCode}>
            <div>
              <div>
                {failedPromotion ? intl.get('promotion-invalid') : ''}
              </div>
              <div className="form-group">
                <div>
                  <input id="promotion_form_code" name="PromotionCode" className="form-control" type="text" defaultValue="" onChange={this.setPromotionCode} />
                </div>
              </div>
            </div>
            <div className="form-group">
              <button id="apply_promotion" className="btn btn-add-promotion" type="submit">
                {intl.get('apply-promotion')}
              </button>
              <button id="cancel_add_promotion" className="btn btn-add-promotion" type="button" onClick={() => { this.closePromotionForm(); }}>
                {intl.get('cancel')}
              </button>
            </div>
          </form>
        </div>
      );
    }
    return (
      <div className="add-promotion-container">
        <button className="btn btn-add-promotion" type="button" id="open_promotion_form_button" onClick={() => { this.openPromotionForm(); }}>
          {intl.get('add-promotion')}
        </button>
      </div>
    );
  }
}
export default AddPromotionContainer;
