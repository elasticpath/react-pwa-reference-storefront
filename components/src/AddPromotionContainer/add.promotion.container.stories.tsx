import React from 'react';
import { storiesOf } from '@storybook/react';
import { MemoryRouter } from 'react-router';

import AddPromotionContainer from "./add.promotion.container";

const cartData = {
    "_order": [
    {
      "_couponinfo": [
        {
          "_couponform": [
            {
              "links": [
                {
                  "rel": "applycouponaction",
                  "type": "coupons.apply-coupon-to-order-form",
                  "uri": "/coupons/orders/vestri/gq2dezrsmrsdgljsmm3wiljuhfrtqllbhfqwmljtgyydcnrsgbqtizrwmu=/form",
                  "href": "http://reference.epdemos.com/cortex/coupons/orders/vestri/gq2dezrsmrsdgljsmm3wiljuhfrtqllbhfqwmljtgyydcnrsgbqtizrwmu=/form"
                }
              ]
            }
          ]
        }
      ]
    }
  ]
};

storiesOf('AddPromotionContainer', module)
  .addDecorator(story => (
    <MemoryRouter initialEntries={['/']}>{story()}</MemoryRouter>
  ))
  .add('AddPromotionContainer', () => <AddPromotionContainer data={cartData}  onSubmittedPromotion={() => {}} />);
