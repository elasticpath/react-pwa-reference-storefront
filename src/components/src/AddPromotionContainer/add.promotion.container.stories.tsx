/**
 * Copyright Â© 2019 Elastic Path Software Inc. All rights reserved.
 *
 * This is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this license. If not, see
 *
 *     https://www.gnu.org/licenses/
 *
 *
 */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { MemoryRouter } from 'react-router';
import { object, text } from '@storybook/addon-knobs/react';
import Readme from './README.md';
import { textToFunc } from '../../../../storybook/utils/storybookUtils';
import AddPromotionContainer from './add.promotion.container';

const cartData = {
  _order: [
    {
      _couponinfo: [
        {
          _couponform: [
            {
              links: [
                {
                  rel: 'applycouponaction',
                  type: 'coupons.apply-coupon-to-order-form',
                  uri: '/coupons/orders/vestri/gq2dezrsmrsdgljsmm3wiljuhfrtqllbhfqwmljtgyydcnrsgbqtizrwmu=/form',
                  href: 'http://reference.epdemos.com/cortex/coupons/orders/vestri/gq2dezrsmrsdgljsmm3wiljuhfrtqllbhfqwmljtgyydcnrsgbqtizrwmu=/form',
                },
              ],
            },
          ],
        },
      ],
    },
  ],
};

storiesOf('Components|AddPromotionContainer', module)
  .addParameters({
    readme: {
      // Show readme at the addons panel
      sidebar: Readme,
    },
  })
  .addDecorator(story => (
    <MemoryRouter initialEntries={['/']}>{story()}</MemoryRouter>
  ))
  .add('AddPromotionContainer', () => {
    let onSubmittedPromotionFuncText = text('onSubmittedPromotion','() => {alert("onSubmitPromotion invoked")}');
    return <AddPromotionContainer data={object('data', cartData)} onSubmittedPromotion={()=>textToFunc(onSubmittedPromotionFuncText)} />
  });
