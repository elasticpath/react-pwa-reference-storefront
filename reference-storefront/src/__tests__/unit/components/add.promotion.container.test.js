/**
 * Copyright Â© 2018 Elastic Path Software Inc. All rights reserved.
 *
 * This is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this license. If not, see
 *
 *     https://www.gnu.org/licenses/
 *
 *
 */

import React from 'react';
import { shallow } from 'enzyme';
import AddPromotionContainer from '../../../components/add.promotion.container';

describe('App', () => {
  it('renders with default props', () => {
    const cartData = {};
    const wrapper = shallow(<AddPromotionContainer data={cartData} onSubmittedPromotion={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('opens form by clicking on button', () => {
    const cartData = {};
    const wrapper = shallow(<AddPromotionContainer data={cartData} onSubmittedPromotion={() => {}} />);
    const button = wrapper.find('button');
    button.simulate('click');
    expect (wrapper.find('form').exists()).toEqual(true);
    wrapper.unmount();
  });
  
  it('submits form', () => {
    const cartData = {};
    const wrapper = shallow(<AddPromotionContainer data={cartData} onSubmittedPromotion={() => {}} />);
    const button = wrapper.find('button');
    wrapper.instance().submitPromotionCode = jest.fn();
    wrapper.update();
    button.simulate('click');
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(wrapper.instance().submitPromotionCode).toBeCalled();
    wrapper.unmount();
  });
});
