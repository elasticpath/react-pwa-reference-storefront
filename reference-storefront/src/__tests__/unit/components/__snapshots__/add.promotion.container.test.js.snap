// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders with default props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddPromotionContainer
    data={Object {}}
    onSubmittedPromotion={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <button
        className="btn btn-add-promotion"
        id="open_promotion_form_button"
        onClick={[Function]}
        type="button"
/>,
      "className": "add-promotion-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
        "className": "btn btn-add-promotion",
        "id": "open_promotion_form_button",
        "onClick": [Function],
        "type": "button",
      },
      "ref": null,
      "rendered": "",
      "type": "button",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <button
          className="btn btn-add-promotion"
          id="open_promotion_form_button"
          onClick={[Function]}
          type="button"
/>,
        "className": "add-promotion-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "btn btn-add-promotion",
          "id": "open_promotion_form_button",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "",
        "type": "button",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
